beaker/application.py

```diff
@@ -1,418 +1,963 @@
                     )
+                assert isinstance(existing_action.action, SubroutineFnWrapper)
+                self.deregister_bare_method(existing_action.action)
+            self.bare_actions[for_action] = OnCompleteAction(
+                action=sub, call_config=call_config
+            )
 
-        self.on_create: Optional[HandlerFunc] = (
-            all_creates.pop() if len(all_creates) == 1 else None
-        )
-        self.on_update: Optional[HandlerFunc] = (
-            all_updates.pop() if len(all_updates) == 1 else None
+    def deregister_bare_method(
+        self,
+        action_name_or_reference: OnCompleteActionName | SubroutineFnWrapper,
+        /,
+    ) -> None:
+        if isinstance(action_name_or_reference, str):
+            self.bare_actions.pop(cast(OnCompleteActionName, action_name_or_reference))
+        else:
+            method = action_name_or_reference
+            _remove_first_match(self.bare_actions, lambda _, v: v.action is method)
+
+    @overload
+    def external(
+        self,
+        fn: HandlerFunc,
+        /,
+    ) -> ABIReturnSubroutine:
+        ...
+
+    @overload
+    def external(
+        self,
+        /,
+        *,
+        method_config: MethodConfig | MethodConfigDict | None = None,
+        name: str | None = None,
+        authorize: SubroutineFnWrapper | None = None,
+        bare: bool | None = False,
+        read_only: bool = False,
+        override: bool | None = False,
+    ) -> DecoratorFuncType:
+        ...
+
+    def external(
+        self,
+        fn: HandlerFunc | None = None,
+        /,
+        *,
+        method_config: MethodConfig | MethodConfigDict | None = None,
+        name: str | None = None,
+        authorize: SubroutineFnWrapper | None = None,
+        bare: bool | None = False,
+        read_only: bool = False,
+        override: bool | None = False,
+    ) -> DecoratorResultType | DecoratorFuncType:
+        """
+        Add the method decorated to be handled as an ABI method for the Application
+
+        Args:
+            fn: The function being wrapped.
+            method_config:  <TODO>
+            name: Name of ABI method. If not set, name of the python method will be used.
+                Useful for method overriding.
+            authorize: a subroutine with input of ``Txn.sender()`` and output uint64
+                interpreted as allowed if the output>0.
+            bare:
+            read_only: Mark a method as callable with no fee using dryrun or simulate
+            override:
+
+        Returns:
+            The original method with additional elements set in it's
+            :code:`__handler_config__` attribute
+        """
+
+        def decorator(func: HandlerFunc) -> SubroutineFnWrapper | ABIReturnSubroutine:
+            sig = inspect.signature(func)
+            nonlocal bare
+            if bare is None:
+                bare = not sig.parameters
```

Comment:
Unless the flag was explicitly set I don't think we should force bare on no args, since the method might be `doit()void` but would still want to match on the method selector instead of on the OnComplete alone

Reply:
Interesting - this was to preserve the previous behaviour of beaker. More than happy to change it to only making a `bare` method on request. Note that this is the case if the user uses `@external` already - `bare=False` is the default. `bare=None` (ie make it bare if it's got no-args) is only used through `@create`, `@opt_in`, et. al.

Reply:
Thinking about this more, this definitely makes sense. Beaker is for building ARC-4 apps. Having to explicitly tell it to make an ABI method rather than a bare method, just because - say for example - your `@opt_in` method takes no args, seems counterintuitive. I'll make that code change and adjust the migration guide to point out this "breaking" change.

Reply:
I've gone ahead and made this change, and it's actually cleaned up the code a bit, and, more importantly, eliminates the need for users to assert/cast certain decorated functions to, for example, use them with `ApplicationClient.call`.